File: java/dsa/data-structures/binary-tree/same-tree.java
/** Definition for a binary tree node */
// public class TreeNode {
//     int val; TreeNode left; TreeNode right;
// }

public boolean isSameTree(TreeNode p, TreeNode q) { // @viz: function_start
  if (p == null && q == null) { // @viz: base_both_null
    return true; // @viz: base_both_null
  }
  if (p == null || q == null || p.val != q.val) { // @viz: base_mismatch
    return false; // @viz: base_mismatch
  }

  boolean left = isSameTree(p.left, q.left); // @viz: recurse_left
  boolean right = isSameTree(p.right, q.right); // @viz: recurse_right

  return left && right; // @viz: return_result
} // @viz: return_result

File: javascript/dsa/data-structures/binary-tree/same-tree.js
/** Definition for a binary tree node */
// class TreeNode {
//     val, left, right
// }

function isSameTree(p, q) { // @viz: function_start
  if (!p && !q) { // @viz: base_both_null
    return true; // @viz: base_both_null
  }
  if (!p || !q || p.val !== q.val) { // @viz: base_mismatch
    return false; // @viz: base_mismatch
  }

  const leftResult = isSameTree(p.left, q.left); // @viz: recurse_left
  const rightResult = isSameTree(p.right, q.right); // @viz: recurse_right

  return leftResult && rightResult; // @viz: return_result
} // @viz: return_result